python/hail/type.py

```diff
@@ -1,12 +1,582 @@
+        jtype = scala_object(Env.hail_package().expr, 'TArray').apply(element_type._jtype)
+        self.element_type = element_type
+        super(TArray, self).__init__(jtype)
+
+    @classmethod
+    def _from_java(cls, jtype):
+        t = TArray.__new__(cls)
+        t.element_type = Type._from_java(jtype.elementType())
+        t._jtype = jtype
+        return t
+
+    def _convert_to_py(self, annotation):
+        if annotation:
+            lst = scala_package_object(Env.hail_package().utils).iterableToArrayList(annotation)
+            return [self.element_type._convert_to_py(x) for x in lst]
+        else:
+            return annotation
+
+    def _convert_to_j(self, annotation):
+        if annotation is not None:
+            return scala_package_object(Env.hail_package().utils).arrayListToISeq(
+                [self.element_type._convert_to_j(elt) for elt in annotation]
+            )
+        else:
+            return annotation
+
+    def _typecheck(self, annotation):
+        if annotation:
+            if not isinstance(annotation, list):
+                raise TypeCheckError("TArray expected type 'list', but found type '%s'" % type(annotation))
+            for elt in annotation:
+                self.element_type._typecheck(elt)
+
+
+class TSet(Type):
+    """
+    Hail type corresponding to set
+
+    :param element_type: type of set elements
+    :type element_type: :class:`.Type`
+
+    :ivar element_type: type of set elements
+    :vartype element_type: :class:`.Type`
+    """
+
+    def __init__(self, element_type):
+        """
+        :param :class:`.Type` element_type: Hail type of set element
+        """
+        jtype = scala_object(Env.hail_package().expr, 'TSet').apply(element_type._jtype)
+        self.element_type = element_type
+        super(TSet, self).__init__(jtype)
+
+    @classmethod
+    def _from_java(cls, jtype):
+        t = TSet.__new__(cls)
+        t.element_type = Type._from_java(jtype.elementType())
+        t._jtype = jtype
+        return t
+
+    def _convert_to_py(self, annotation):
+        if annotation:
+            lst = scala_package_object(Env.hail_package().utils).iterableToArrayList(annotation)
+            return set([self.element_type._convert_to_py(x) for x in lst])
+        else:
+            return annotation
+
+    def _convert_to_j(self, annotation):
+        if annotation is not None:
+            return scala_package_object(Env.hail_package().utils).arrayListToSet(
+                [self.element_type._convert_to_j(elt) for elt in annotation]
+            )
+        else:
+            return annotation
+
+    def _typecheck(self, annotation):
+        if annotation:
+            if not isinstance(annotation, set):
+                raise TypeCheckError("TSet expected type 'set', but found type '%s'" % type(annotation))
+            for elt in annotation:
+                self.element_type._typecheck(elt)
+
+
+class TDict(Type):
+    """
+    Hail type corresponding to dict
```

Comment:
Python dicts can have any key type.  Mention that Hail values have/require string keys.